op-node/rollup/derive/system_config.go

```diff
@@ -69,50 +72,86 @@ func ProcessSystemConfigUpdateLogEvent(destSysCfg *eth.SystemConfig, ev *types.L
 	}
 	// indexed 1
 	updateType := ev.Topics[2]
-	// unindexed data
+
+	// Create a reader of the unindexed data
+	reader := bytes.NewReader(ev.Data)
+
+	// Helper function to prevent code duplication.
+	readWord := func() (b []byte) {
+		if _, err := reader.Read(b); err != nil {
+			// The possible error returned by `Read` is ignored due to the length check of the unindexed data in
+			// all cases of the below switch statement. While we don't panic here, this log should *never* be emitted.
+			logger.Crit("failed to read word from unindexed log data")
```

Comment:
I don't love crashing on malicious input. The L1 multisig secures the abi of the data coming in here, really bad things can happen in the L1 multisig is compromised. Ideally a compromised multisig cannot crash nodes on the network as well as not steal any money from the bridge

Reply:
The previous impl did crash on malicious input as well, would it be difficult to handle malicious input at part of this PR

Reply:
We have a design doc about safety vs liveness w.r.t. L1 receipts that we read. If one of these contracts is off, something is already severely wrong, and we should likely halt.

Reply:
`logger.Crit` calls `panic` right? We could halt without crashing the software? Sounds like that is too large of a change for this PR then?

Reply:
The derivation pipeline has a `NewCriticalError(...)` typed error you can return to halt without crashing. But it's a bit difficult because of the function not returning it.

Ideally we check the input length once, and then do the reading safely without unnecessary and verbose error handling.

Reply:
Due to the `readWord` reading 32 bytes every time + the length checks in each of the switch cases, this log should be impossible to hit regardless of a malicious or honest input, no? I mainly put it there to protect against any changes in this codepath that don't assert adequate length before reading.

Reply:
Possible solution: track how much we've read from the input in the helper func, return zeroed-data if the reader goes out of bounds, and after reading everything check if we have consumed everything. So then there's only a single error to check

Reply:
I'm a fan of this route. @tynes ?

Reply:
This sounds good to me! Good thinking @protolambda

Reply:
I think a way to do this would be implementing a custom type like `WordReader`, this would let us get rid of the closure and would potentially be useful elsewhere. It could have public methods `Check` for the assertion as well as implement `io.Reader` and only read out `[32]byte` as slices to fit the interface, could also add `ReadHash() common.Hash` for simplicity

Reply:
Made the changes- sorry for the delay, had to step away for a sec. Should we add a test that asserts `ProcessSystemConfigUpdateLogEvent` properly errors out when given an unexpected # of bytes in the log?

Reply:
> Made the changes- sorry for the delay, had to step away for a sec. Should we add a test that asserts `ProcessSystemConfigUpdateLogEvent` properly errors out when given an unexpected # of bytes in the log?

RFC @protolambda (sorry, know you've already approved) ^

There are tests that assert this function's correctness, but given how critical it is, would it be a good idea to add a negative case here? Can also put it up in another PR if that'd be better.

Reply:
These look like good changes to me, and it's not less critical/complex than previously, so I'm fine to merge it. If you see opportunity to implement the test then it's welcome ofc.

Reply:
Mergify will pick it up if you mark these comments as resolved.